`No repository found for plugin golangci-lint` in 2024.11.3 (when run with custom `MISE_DATA_DIR`?)

**Describe the bug**

2024.11.3 broke golangci-lint plugin:

```
DEBUG ARGS: /opt/homebrew/bin/mise plugin install --yes -q golangci-lint
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/go-junit-report"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/skaffold"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/staticcheck"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/kube-code-generator"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/gotestsum"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/kube-controller-tools"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/kustomize"
DEBUG vfox_plugin: "/Users/XXXX/code_/PROJECT/bin/plugins/setup-envtest"
DEBUG config: ~/code_/PROJECT/.mise.toml
DEBUG config: ~/.config/mise/config.toml
Error:
   0: No repository found for plugin golangci-lint

Location:
   src/plugins/asdf_plugin.rs:78

Version:
   2024.11.3 macos-arm64 (2024-11-06)

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
make[1]: *** [mise-plugin-install] Error 1
make: *** [Makefile:91: golangci-lint.download] Error 2
```

**To Reproduce**

Run `mise plugin install --yes golangci-lint` with a custom `MISE_DATA_DIR`, e.g.

```
RUST_BACKTRACE=1 MISE_VERBOSE=1 MISE_DATA_DIR=./bin mise plugin install --yes golangci-lint
DEBUG ARGS: mise plugin install --yes golangci-lint
DEBUG vfox_plugin: "./bin/plugins/go-junit-report"
DEBUG vfox_plugin: "./bin/plugins/skaffold"
DEBUG vfox_plugin: "./bin/plugins/staticcheck"
DEBUG vfox_plugin: "./bin/plugins/kube-code-generator"
DEBUG vfox_plugin: "./bin/plugins/gotestsum"
DEBUG vfox_plugin: "./bin/plugins/kube-controller-tools"
DEBUG vfox_plugin: "./bin/plugins/kustomize"
DEBUG vfox_plugin: "./bin/plugins/setup-envtest"
DEBUG config: ~/code_/kubernetes-ingress-controller/.mise.toml
DEBUG config: ~/.config/mise/config.toml
Error:
   0: No repository found for plugin golangci-lint

Location:
   src/plugins/asdf_plugin.rs:78

Version:
   2024.11.3 macos-arm64 (2024-11-06)

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 4 frames hidden ⋮
   5: mise::plugins::asdf_plugin::AsdfPlugin::get_repo_url::h186d3bae09e887d9
      at <unknown source file>:<unknown line>
   6: <mise::plugins::asdf_plugin::AsdfPlugin as mise::plugins::Plugin>::install::h8efd5b81852e5e9b
      at <unknown source file>:<unknown line>
   7: <mise::plugins::asdf_plugin::AsdfPlugin as mise::plugins::Plugin>::ensure_installed::h078f9f499b3df12c
      at <unknown source file>:<unknown line>
   8: mise::cli::plugins::install::PluginsInstall::install_one::hfb9e4a70c140ec3a
      at <unknown source file>:<unknown line>
   9: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend::h0a305d511bc2cdd9
      at <unknown source file>:<unknown line>
  10: rayon::iter::plumbing::bridge_producer_consumer::helper::hfe0406dff11fa874
      at <unknown source file>:<unknown line>
  11: <rayon::vec::IntoIter<T> as rayon::iter::IndexedParallelIterator>::with_producer::he977c2ec67a2ffff
      at <unknown source file>:<unknown line>
  12: rayon::iter::extend::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend::hb8bacfab8c921b40
      at <unknown source file>:<unknown line>
  13: rayon::result::<impl rayon::iter::FromParallelIterator<core::result::Result<T,E>> for core::result::Result<C,E>>::from_par_iter::h9d9ebf53b7378197
      at <unknown source file>:<unknown line>
  14: <rayon_core::job::StackJob<L,F,R> as rayon_core::job::Job>::execute::h810058c34bce5d09
      at <unknown source file>:<unknown line>
  15: rayon_core::registry::WorkerThread::wait_until_cold::hb8d80409e54f2d72
      at <unknown source file>:<unknown line>
  16: rayon_core::registry::ThreadBuilder::run::ha798e54870547d13
      at <unknown source file>:<unknown line>
  17: std::sys::backtrace::__rust_begin_short_backtrace::hab24813a0d014675
      at <unknown source file>:<unknown line>
  18: core::ops::function::FnOnce::call_once{{vtable.shim}}::hd369b589a0def31b
      at <unknown source file>:<unknown line>
  19: std::sys::pal::unix::thread::Thread::new::thread_start::h9e79a8d54639c27f
      at <unknown source file>:<unknown line>
  20: __pthread_deallocate<unknown>
      at <unknown source file>:<unknown line>

Run with COLORBT_SHOW_HIDDEN=1 environment variable to disable frame filtering.
Run with RUST_BACKTRACE=full to include source snippets.
```

**Expected behavior**

No error

**`mise doctor` output**

```text
version: 2024.11.3 macos-arm64 (2024-11-06)
activated: no
shims_on_path: no

build_info:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS
  Built: Wed, 6 Nov 2024 11:51:34 +0000
  Rust Version: rustc 1.82.0 (f6e511eec 2024-10-15) (Homebrew)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (arm64-apple-darwin24.0)

dirs:
  data: ~/.local/share/mise
  config: ~/.config/mise
  cache: ~/Library/Caches/mise
  state: ~/.local/state/mise
  shims: ~/.local/share/mise/shims

config_files:
  ~/.config/mise/config.toml
  ~/code_/kubernetes-ingress-controller/.mise.toml

backends:
  cargo
  core
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:
  golangci-lint          ssh://git@github.com/hypnoglow/asdf-golangci-lint.git#fee877b
  kube-code-generator    ssh://git@github.com/pmalek/asdf-kube-code-generator.git#4edd5bd
  kube-controller-tools  ssh://git@github.com/jimmidyson/asdf-kube-controller-tools.git#37526dc
  kustomize              ssh://git@github.com/Banno/asdf-kustomize.git#8e929af

toolset:

env_vars:
  (none)

settings:
  activate_aggressive = false
  all_compile = false
  always_keep_download = false
  always_keep_install = false
  asdf_compat = false
  cache_prune_age = "30d"
  ci = false
  color = true
  debug = false
  disable_backends = []
  disable_default_registry = false
  disable_hints = []
  disable_tools = []
  experimental = true
  fetch_remote_versions_cache = "1h"
  fetch_remote_versions_timeout = "10s"
  go_default_packages_file = "~/.default-go-packages"
  go_download_mirror = "https://dl.google.com/go"
  go_repo = "https://github.com/golang/go"
  go_set_gopath = false
  go_set_goroot = true
  go_skip_checksum = false
  http_timeout = "30s"
  jobs = 4
  legacy_version_file = true
  legacy_version_file_disable_tools = []
  libgit2 = true
  lockfile = false
  log_level = "info"
  not_found_auto_install = true
  paranoid = false
  pin = false
  plugin_autoupdate_last_check_duration = "7d"
  quiet = false
  raw = false
  trace = false
  trusted_config_paths = []
  use_versions_host = true
  verbose = false
  yes = false

  [cargo]
  binstall = true

  [node]

  [pipx]
  uvx = false

  [python]
  default_packages_file = "~/.default-python-packages"
  pyenv_repo = "https://github.com/pyenv/pyenv.git"
  venv_auto_create = false
  venv_stdlib = false

  [ruby]
  default_packages_file = "~/.default-gems"
  ruby_build_repo = "https://github.com/rbenv/ruby-build.git"
  ruby_install = false
  ruby_install_repo = "https://github.com/postmodern/ruby-install.git"

  [status]
  missing_tools = "if_other_versions_installed"
  show_env = false
  show_tools = false
No warnings found
1 problem found:

1. mise is not activated, run mise help activate or
   read documentation at https://mise.jdx.dev for activation instructions.
   Alternatively, add the shims directory ~/.local/share/mise/shims to PATH.
   Using the shims directory is preferred for non-interactive setups.
```

**Additional context**

Not specifying `MISE_DATA_DIR` seems to work just fine:

```
mise plugin install --yes --force golangci-lint
mise WARN  git clone failed: authentication required but no callback set; class=Ssh (23); code=Auth (-16)
mise plugin:golangci-lint ✓ ssh://git@github.com/hypnoglow/asdf-golangci-lint.git#fee877b
```

Link to failing CI: https://github.com/Kong/kubernetes-ingress-controller/actions/runs/11706610991/job/32604256116?pr=6625#step:5:99

--- a/src/cli/plugins/ls.rs
+++ b/src/cli/plugins/ls.rs
@@ -48,10 +48,12 @@ impl PluginsLs {
         let mut tools = plugins::list2()?;
 
         if self.all {
-            for (plugin, full) in config.get_shorthands() {
-                let mut ep = AsdfPlugin::new(plugin.to_string());
-                ep.repo_url = Some(full_to_url(full));
-                tools.insert(ep.name.clone(), Box::new(ep));
+            for (plugin, backends) in config.get_shorthands() {
+                for full in backends {
+                    let mut ep = AsdfPlugin::new(plugin.to_string());
+                    ep.repo_url = Some(full_to_url(full));
+                    tools.insert(ep.name.clone(), Box::new(ep));
+                }
             }
         } else if self.user && self.core {
         } else if self.core {

--- a/src/cli/plugins/ls_remote.rs
+++ b/src/cli/plugins/ls_remote.rs
@@ -41,15 +41,17 @@ impl PluginsLsRemote {
             warn!("default shorthands are disabled");
         }
 
-        for (plugin, repo) in shorthands {
-            let installed = if !self.only_names && installed_plugins.contains(plugin.as_str()) {
-                "*"
-            } else {
-                " "
-            };
-            let url = if self.urls { repo } else { "" };
-            let plugin = pad_str(plugin, max_plugin_len, Alignment::Left, None);
-            miseprintln!("{} {}{}", plugin, installed, url);
+        for (plugin, backends) in shorthands {
+            for repo in backends {
+                let installed = if !self.only_names && installed_plugins.contains(plugin.as_str()) {
+                    "*"
+                } else {
+                    " "
+                };
+                let url = if self.urls { repo } else { "" };
+                let plugin = pad_str(plugin, max_plugin_len, Alignment::Left, None);
+                miseprintln!("{} {}{}", plugin, installed, url);
+            }
         }
 
         Ok(())

--- a/src/config/mod.rs
+++ b/src/config/mod.rs
@@ -47,7 +47,7 @@ pub struct Config {
     env_with_sources: OnceCell<EnvWithSources>,
     all_aliases: OnceLock<AliasMap>,
     repo_urls: HashMap<String, String>,
-    shorthands: OnceLock<HashMap<String, String>>,
+    shorthands: OnceLock<Shorthands>,
     tasks: OnceCell<BTreeMap<String, Task>>,
     tool_request_set: OnceCell<ToolRequestSet>,
 }
@@ -160,7 +160,7 @@ impl Config {
             None => self
                 .get_shorthands()
                 .get(plugin_name)
-                .map(|full| registry::full_to_url(full)),
+                .map(|full| registry::full_to_url(&full[0])),
         }
     }
 

--- a/src/shorthands.rs
+++ b/src/shorthands.rs
@@ -2,13 +2,14 @@ use std::collections::HashMap;
 use std::path::PathBuf;
 
 use eyre::Result;
+use itertools::Itertools;
 use toml::Table;
 
 use crate::config::Settings;
 use crate::registry::REGISTRY;
 use crate::{dirs, file};
 
-pub type Shorthands = HashMap<String, String>;
+pub type Shorthands = HashMap<String, Vec<String>>;
 
 pub fn get_shorthands(settings: &Settings) -> Shorthands {
     let mut shorthands = HashMap::new();
@@ -16,11 +17,16 @@ pub fn get_shorthands(settings: &Settings) -> Shorthands {
         shorthands.extend(
             REGISTRY
                 .iter()
-                .filter(|(_, full)| {
-                    let backend = &full[0].split(':').next().unwrap().to_string();
-                    backend == "asdf" || backend == "vfox"
+                .map(|(id, full)| {
+                    (
+                        id.to_string(),
+                        full.iter()
+                            .filter(|f| f.starts_with("asdf:") || f.starts_with("vfox:"))
+                            .map(|f| f.to_string())
+                            .collect_vec(),
+                    )
                 })
-                .map(|(k, v)| (k.to_string(), v[0].to_string())),
+                .filter(|(_, fulls)| !fulls.is_empty()),
         );
     };
     if let Some(f) = &settings.shorthands_file {
@@ -46,7 +52,7 @@ fn parse_shorthands_file(mut f: PathBuf) -> Result<Shorthands> {
     let mut shorthands = HashMap::new();
     for (k, v) in toml {
         if let Some(v) = v.as_str() {
-            shorthands.insert(k, v.to_string());
+            shorthands.insert(k, vec![v.to_string()]);
         }
     }
     Ok(shorthands)
@@ -70,9 +76,9 @@ mod tests {
         let mut settings = Settings::get().deref().clone();
         settings.shorthands_file = Some("../fixtures/shorthands.toml".into());
         let shorthands = get_shorthands(&settings);
-        assert_str_eq!(shorthands["elixir"], "asdf:mise-plugins/mise-elixir");
-        assert_str_eq!(shorthands["node"], "https://node");
-        assert_str_eq!(shorthands["xxxxxx"], "https://xxxxxx");
+        assert_str_eq!(shorthands["elixir"][0], "asdf:mise-plugins/mise-elixir");
+        assert_str_eq!(shorthands["node"][0], "https://node");
+        assert_str_eq!(shorthands["xxxxxx"][0], "https://xxxxxx");
     }
 
     #[test]

